src/peergos/server/net/GatewayHandler.java

```diff
@@ -0,0 +1,108 @@
+package peergos.server.net;
+
+import com.sun.net.httpserver.*;
+import peergos.server.util.Logging;
+import peergos.server.util.*;
+import peergos.shared.*;
+import peergos.shared.user.*;
+import peergos.shared.user.fs.*;
+import peergos.shared.util.*;
+
+import java.nio.file.*;
+import java.util.*;
+import java.util.logging.*;
+
+public class GatewayHandler implements HttpHandler {
+	private static final Logger LOG = Logging.LOG();
+	private static final boolean LOGGING = true;
+
+    private final String domainSuffix;
+    private final NetworkAccess network;
+    private final Crypto crypto;
+
+    public GatewayHandler(String domainSuffix, Crypto crypto, NetworkAccess network) {
+        this.domainSuffix = domainSuffix;
+        this.crypto = crypto;
+        this.network = network;
+    }
+
+    @Override
+    public void handle(HttpExchange httpExchange) {
+        long t1 = System.currentTimeMillis();
+        String path = httpExchange.getRequestURI().getPath();
+        try {
+            path = path.substring(1).replaceAll("//", "/");
+            if (path.length() == 0)
+                path = "index.html";
+
+            String domain = httpExchange.getRequestHeaders().getFirst("Host");
+            if (! domain.endsWith(domainSuffix))
+                throw new IllegalStateException("Incorrect domain! " + domain);
+            String owner = domain.substring(0, domain.length() - domainSuffix.length());
+            Path toProfileEntry = Paths.get(owner).resolve(".profile").resolve("webroot");
+            AbsoluteCapability capToWebrootFiled = UserContext.getPublicCapability(toProfileEntry, network).join();
+            Optional<FileWrapper> profileField = network.getFile(capToWebrootFiled, owner).join();
+            Path toWebroot = Paths.get(new String(Serialize.readFully(profileField.get(), crypto, network).join()));
+
+            AbsoluteCapability capToWebroot = UserContext.getPublicCapability(toWebroot, network).join();
+            Optional<FileWrapper> webroot = network.getFile(capToWebroot, owner).join();
+            Optional<FileWrapper> assetOpt = webroot.get().getDescendentByPath(path, crypto.hasher, network).join();
+            if (assetOpt.isEmpty()) {
+                httpExchange.getResponseHeaders().set("Content-Type", "text/plain");
+                httpExchange.sendResponseHeaders(404, 0);
+                return;
+            }
+            FileWrapper file = assetOpt.get();
+            if (file.isDirectory())
+                throw new IllegalStateException("Directory listings not allowed");
```

Comment:
HTTP error code in a response?

Reply:
I've made it 404 which is what github does